### PR TITLE
bash script to create new, bare app

### PR DESCRIPTION
**Pull request by rwcarlsen**
How about a much smaller step than something like #8252 - a script for generating new apps.  New moose users just ``$MOOSE_DIR/scripts/newapp.sh MyNewAppName && cd mynewappname && make -jn`` to get started with a new app.  github, forking, and git become (highly recommended) optional tasks.  Another thing to consider would be moving all the new-app/stork assets into the moose repository - eliminating the maintenance burden of an extra repo.

### CODE REVIEW

**Comment by friedmud** [comment]
Yep.  The "maintenance burden" of 61 commits in _3 years_.  In 2016 alone there was the enormous amount of _10_!!

....

Joking aside: I don't see how moving it into the repo is useful.  If we don't want to do the forking let's just change the instructions to a clone.

If we don't want the clone to be tied back to Stork then have the script remove the `.git` directory.  We could then provide instructions for how they could create a new repo on GitHub and push that new App up into it if they want.

**Comment by permcody** [comment]
I'm not opposed to the existence of this script. 

The question is, do we redo our "Getting Started" instructions to allow for both methods? Do we drop the fork stork stuff? I'd rather keep that around for some people, but this script is handy for other people. However there are risks in having too many ways of accomplishing the same thing.

**Comment by rwcarlsen** [comment]
I agree with @permcody on the "too many ways" problem.  We should decide and focus on one way.  I think something like this combined with a move of stork assets into a moose repo subdir and updates to getting started docs would be best.  If we did that, stork should stay around (deprecated) for a while anyway.  If we don't point to something like this for new users, then it's probably not worth having yet another script.

**Comment by WilkAndy** [comment]
I don't like bash scripts.  They often seem to fail on my crazy architectures.   Probably this one will be fine, but perhaps a python version is preferable?

**Comment by friedmud** [comment]
Ok - I've thought more about this - and I'm ok with putting the prototype App inside the repo and using a script to copy it out and fix it up to create a new app.

The reason for the change of heart is that this will allow you to have a "static-ish" version of MOOSE that you can keep creating new Apps from... and those new Apps will always work with your version of MOOSE.  Right now... if you have an older copy of MOOSE and you clone Stork there's a possibility that the two things will be out of sync and you'll get compile errors (or other kinds of errors).  Keeping them together fixes that.

I also like that creating a new app is all "local"... so you can do it as often as you like - even if you're disconnected from the net.

As for Bash vs. Python: I don't have a strong opinion.  To me: this is Bash kind of work (and the script here is fine).  There isn't much logic or looping or anything else that really warrants Python.  But I'm fine with whatever.

One request though: I think the new App directory should be set up as a local git repo... and the first commit should either be completed or staged just like it is now with Stork.  We can also print out the next few commands / instructions they would need to push the repo up to GitHub as a not-so-subtle hint :-)

**Comment by rwcarlsen** [comment]
Great!  I'll get to work on moving stork assets and modifying the script in
the coming days.

**Comment by aeslaughter** [comment]
Personally, I think we should use jinja2 templates via python. I am using it for the website, including latex templates, and it works quite nicely. But, it won't bother me if we do it another way.

**Comment by rwcarlsen** [comment]
This should be about ready.  We now have a prototype in-moose stork civet recipe.  As soon as this gets merged, we can merge that recipe into external app tests (or wherever) to always run by default.  I also added git staging.  After this gets merged, I'll work on updating docs/sites, etc. for deprecating the stork repository.

File: scripts/newmoose.sh
Comment (permcody) [inline scripts/newmoose.sh]: This name doesn't match the name of the script.

File: scripts/newmoose.sh
Comment (permcody) [inline scripts/newmoose.sh]: I prefer "CamelCase" instead.

File: scripts/newmoose.sh
Comment (permcody) [inline scripts/newmoose.sh]: Umm - I had MOOSE_DIR defined so this put the new animal under my MOOSE_DIR. This needs to be fixed.

File: scripts/newmoose.sh
Comment (permcody) [inline scripts/newmoose.sh]: $dsnname was CamelCase when I tried this so the "cd" command had caps in it. However, the actual directory it created did not, it was all lowercase.

File: stork/src/base/StorkApp.C.module
Comment (permcody) [inline stork/src/base/StorkApp.C.module]: This file was left over by the script. It should be deleted when I create an app.

**Review by permcody — CHANGES_REQUESTED** [review]
This didn't work for me just yet. I'll test again when you fix up these comments.

File: scripts/newmoose.sh
Comment (rwcarlsen) [inline scripts/newmoose.sh]: It just puts it in the current working dir.  I added a check to print an error when you run the script *inside* the main moose directory.

File: stork/src/base/StorkApp.C.module
Comment (rwcarlsen) [inline stork/src/base/StorkApp.C.module]: fixed

**Comment by rwcarlsen** [comment]
Should be ready for a final look.  Do we really want to prevent users from running the command inside the moose repo - there isn't really anything broken about doing it that way.  And they could also easily ``mv`` the app directory to anywhere else easily.  I don't think it will be any easier for users to have to manually specify the path (they would still need to understand bash path semantics well for that - and then running the script from a separate dir wouldn't be a problem anyway).  It might be nice to put the script in a directory intended to be appended to the user's PATH (e.g. $MOOSE_DIR/bin).

File: scripts/newmoose.sh
Comment (friedmud) [inline scripts/newmoose.sh]: Can we call this `stork.sh`?  I really do like the Stork _name_ at least :-)

File: scripts/newmoose.sh
Comment (friedmud) [inline scripts/newmoose.sh]: There needs to be more checking for `MOOSE_DIR` here... like more verification that it actually points at MOOSE.  I'm worried that the script might do some crazy junk if `MOOSE_DIR` were wrong!

File: scripts/newmoose.sh
Comment (friedmud) [inline scripts/newmoose.sh]: We need to also check that the `modules` directory is writeable and issue a good error in the case that it's not.

More and more I'm seeing MOOSE "installed" by system admins in areas that aren't writeable by users.

File: scripts/newmoose.sh
Comment (friedmud) [inline scripts/newmoose.sh]: I do think it's a good idea to error out here.  We don't want people making apps in the middle of the MOOSE repo.  It will make version control REALLY screwy for them.

File: scripts/newmoose.sh
Comment (friedmud) [inline scripts/newmoose.sh]: Good god this makes me nervous.  Can we put some extra checking on here that `$dir` is NOT their home directory or something?  I don't know how it would get in that state... but we need to be extra careful about these kinds of things.

File: scripts/newmoose.sh
Comment (friedmud) [inline scripts/newmoose.sh]: I like the GitHub instructions!

**Review by friedmud — COMMENTED** [review]
Looking good

File: scripts/newmoose.sh
Comment (rwcarlsen) [inline scripts/newmoose.sh]: ``$dir`` - which is the only place we rm stuff from is a directory that is created by the script.  We also check that $dir doesn't exist before we create it.  I'll remove the ``-r`` flag though for good measure.

File: scripts/newmoose.sh
Comment (rwcarlsen) [inline scripts/newmoose.sh]: Especially since this is most important for new users, it seems like the most (boring) appropriate approach would be to have something descriptive.  I don't have a strong opinion though and am definitely open to suggestions from everyone.

File: scripts/newmoose.sh
Comment (rwcarlsen) [inline scripts/newmoose.sh]: I added an extra check for ``$MOOSE_DIR/framework/include/kernels/KernelBase.h``.

File: scripts/newmoose.sh
Comment (aeslaughter) [inline scripts/newmoose.sh]: Does the script to a 'git init' for you?

**Comment by aeslaughter** [comment]
@friedmud is correct, it should be called 'stork.sh'.

**Comment by aeslaughter** [comment]
I am going to open up a can of worms that I didn't think about until now. I while back we discussed that new applications should come with moose as a submodule, I started working on it but never completed the task for a number of reasons. Is this still something we want to consider?

If we want this do be the default, then the 'stork.sh' inside of MOOSE is problematic. Because you would have to clone moose, run stork, and then not use the moose in which they cloned originally. Just something to consider.

**Comment by permcody** [comment]
That's a really good point @aeslaughter and could sink this whole effort. I was initially against the MOOSE submodule but I'm actually much more for it these days. I think most users are using MOOSE as a submodule these days so cloning MOOSE is less necessary. It doesn't make a lot of sense to clone MOOSE to build stork only to delete MOOSE again.

**Comment by aeslaughter** [comment]
I don't think that this effort is for nothing if we want to go the submodule route, this script will just need to not live inside MOOSE. Could we consider having the first step of MOOSE be run this script, we could maybe do one of those one liners that download and run the script.

**Comment by permcody** [comment]
Well we already had a working script in Python...

**Comment by rwcarlsen** [comment]
There are definitely advantages to both approaches.

Submodule good:
* app creators can easily pin their app to a specific version of moose for themselves and all their app users
* implicit version control of app

Submodule bad:
* users will quickly run into quirks of submodules and git generally and will be required to navigate them from the get-go
* Stork assets live outside MOOSE repo - and maintenance/synchronization issues will exist
* mutliple apps in an environment can't share a moose build/install
* network connection needed to create new app (unless you are rather savvy accomplishing it with a current app-install that has the moose submodule pulled down.

in-moose script good:
* No git quirkiness to navigate
* stork assets live with and are always in sync with moose.
* multiple apps can share a MOOSE build
* no network for new apps - just need moose repo.

in-moose script bad:
* no implicit pinning of app to specific moose version for users
* no implicit version control of app
* users need to be able to run the script correctly

Submodules still trip me up not infrequently FWIW.

**Comment by aeslaughter** [comment]
I think I prefer the script in moose. The reason we don't like repository
based stork is that it was too confusing for new users that have no concept
of version control, which is the majority. Starting with moose as a
submodule will continue to require us to try to explain the intricacies of
version control, including submodules.

It is easy to setup the submodule after the fact and not something that you
really need to worry about until you are thinking of deploying your code,
if people are at that point they better be comfortable enough with moose
and git to be able to set it up, especially if we have some simple
instructions.

File: scripts/newmoose.sh
Comment (rwcarlsen) [inline scripts/newmoose.sh]: yes

File: scripts/newmoose.sh
Comment (rwcarlsen) [inline scripts/newmoose.sh]: done

**Comment by friedmud** [comment]
Man - I can see both ways.

Since I'm currently mostly a "user" I have to say that it drives me nuts that I don't have MOOSE as a submodule of my Stork-based App.  I had a breakage in my code a couple of months ago that was subtle... and required me to go back through both MOOSE and libMesh versions as I went back through my application versions trying to find when it last worked so I could track down the problem.  It was an enormous hassle trying to find libMesh/MOOSE/app versions that worked together and I was beating myself in the head for not having submodules setup.

I like @rwcarlsen 's list... let me make comments based on his thoughts (these aren't rebuttals, just things I was thinking as I read his list):

> app creators can easily pin their app to a specific version of moose for themselves and all their app users

This also can come at a price... many won't ever bother updating MOOSE....

> users will quickly run into quirks of submodules and git generally and will be required to navigate them from the git-go

As long as they don't modify anything in MOOSE it should be ok.  But I agree that submodules can be tricky - especially for someone new!

> Stork assets live outside MOOSE repo - and maintenance/synchronization issues will exist

Like I mentioned before: Stork is not a maintenance problem currently.

> mutliple apps in an environment can't share a moose build/install

This not at all an issue for most users.  Most users only ever have one app.  The ones that have multiple are advanced anyway and can still set MOOSE_DIR to point to an external MOOSE (just like we all do at INL).

> network connection needed to create new app (unless you are rather savvy accomplishing it with a current app-install that has the moose submodule pulled down.

I don't really think this is a big deal.  Most users only create one app... and then they have it for years.  Requiring a network connection for that isn't a problem.

> in-moose script good:
> No git quirkiness to navigate

I agree with this - it does simplify things

> stork assets live with and are always in sync with moose.

Ok - but not a huge deal.

> multiple apps can share a MOOSE build

Yes - but doesn't impact most users

> no network for new apps - just need moose repo.

True - but not a huge win

> in-moose script bad:
> no implicit pinning of app to specific moose version for users

Could be big for users.  If they want a coworker to start helping them they now need to coordinate MOOSE versions somehow.  Or, if they need to go back to an old version of their app they have to try to guess at old versions of MOOSE / libMesh that work with that old version of their app.  It's a damn quagmire.

> no implicit version control of app

Yes and no - you do have the instructions there to help them get started on GitHub

> users need to be able to run the script correctly

Oh man - how little do we think of our users 😉 

> Submodules still trip me up not infrequently FWIW.

I don't really have any trouble with them... but I never work _in_ them.  Submodules are "read only" for me.

**Comment by friedmud** [comment]
Let me say one more thing that could be viewed as a vote for Stork having a MOOSE submodule:

I think we could rethink the way users _think_ (lots of "thinking" going on here :-) about MOOSE.

Envision this: what if a new user's first step toward using MOOSE is to _clone Stork_!  That means that, instead of thinking about "getting MOOSE setup" and then thinking about "mucking around with Examples and Modules and shit in MOOSE" (like everyone inexplicably does instead of creating an app)... we could turn that on its head and make user's _apps_ the center of attention.

We could then make it very clear that their App is gaining all of the functionality from MOOSE (including the modules, etc.) and we could get them down the path of creating their own app more quickly.

The "make_new_application.sh" script (or whatever) would actually clone down the MOOSE submodule and run the libmesh script, and build moose_test and run tests, etc.  All of the stuff we tell people to do now.

Now: You know that I'm not normally for this level of automation... and there's a good possibility I'm going to come back tomorrow and say that this is bullshit.  But I am kind of liking the idea that users get their app... and MOOSE is a bit more of a black box... and then all of our instructions from that point forward are about _adding_ things to their apps (instead of having them muck around in MOOSE).

What do you think?

**Comment by friedmud** [comment]
Oh - one more thing: if each App has its own MOOSE submodule... that means that you can have multiple Apps with _different_ MOOSE versions on your system and everything works correctly.

There are other systems that work this way.  In the Javascript world this is _exactly_ now NodeJS modules work.  All of the dependencies of your script are "installed" _locally_... right into the same directory where your script is!  That means that each script can have a completely different set of modules and module versions its using.

This is as opposed to the Python way of doing things where everything is installed in the system... and then we need complicated "environment managers" like Conda in order to swap out our whole system when we change from working on one project to the next because they need different versions of Python dependencies.

I like the NodeJS way better... and it corresponds to having a MOOSE submodule for each App.

**Comment by rwcarlsen** [comment]
It's not just Node - ruby gems, julia, go, rust's cargo, and many more all vendor dependencies into a single self-contained project.  I don't feel strongly either way.  We still require the users to install a bunch of external dependencies anyway (git, petsc, etc.).  If cloning/building moose separately is such a burden, we could just have users ``curl | bash`` a script that sets up dependencies, etc. for them and optionally appends ``$MOOSE_DIR/bin`` to their path.  Then they could just run the stork script.  The user experience can be made very similar regardless of the approach.  So I think we should decide based on other factors.

**Comment by friedmud** [comment]
Just FYI: Julia _doesn't_ work that way.  It's essentially the same as Python, but no one has built an "environment manager" yet.  Which is a PITA.  I think it's one of the mistakes they've made over there.

For me: the biggest driving factor is that there is no way a user can reliably go back to an older version of their code without the use of submodules.  That right there is a HUGE vote for going with submodules.

**Comment by rwcarlsen** [comment]
Tagged semver releases of MOOSE would certainly aid users in managing
version histories.

**Comment by friedmud** [comment]
@rwcarlsen releases of tightly integrated computational science software have their own issues.  See:

Integration Strategies for Computational Science and Engineering Software. 2009-0655 C, Second International Workshop and Software Engineering for Computational Science & Engineering, Vancouver, Canada, May 23, 2009

And:

http://openresearchsoftware.metajnl.com/articles/10.5334/jors.as/
http://openresearchsoftware.metajnl.com/articles/10.5334/jors.bx/

This is not a time to debate our release or integration strategies.  The question is squarely focused on Stork.

I think I'm ok with this current PR for now... and foregoing the MOOSE submodule.  Even though I think a submodule is a good idea... it's just too much for new users.  We should maybe publish a guide on how to add and use a MOOSE submodule for people who are interested in that though.

**Comment by rwcarlsen** [comment]
Your right - releases is its own other can of worms.

**Comment by rwcarlsen** [comment]
@permcody - merge or close?

**Comment by friedmud** [comment]
Can I request tabling of this until I get to try it out?

**Comment by permcody** [comment]
That and we need to make sure the "Getting Started" documentation on mooseframework is up to date. That's critical for this kind of merge.

**Comment by rwcarlsen** [comment]
We don't want to update the docs until after a merge right? Merging doesn't
break the old way, but updated docs wouldn't work unless this were already
merged.

**Comment by rwcarlsen** [comment]
ping?

**Comment by permcody** [comment]
@friedmud - If you have any issues with this PR, let us know. We want to get training type materials cleaned up before the next workshop. I'll test this today and merge unless you'd like to independently test.

**Comment by friedmud** [comment]
Looking at it

**Comment by rwcarlsen** [comment]
Rebased and updated stork assets to use googletest for unit tests.

**Comment by rwcarlsen** [comment]
ping?

**Comment by aeslaughter** [comment]
I am for this change, but we need to update the documentation on the current mooseframework.org. I would like the getting started page to just include running this script in the list of instructions.
1. Setup MOOSE environment.
2. Clone MOOSE.
3. Compile libmesh.
4. Compile MOOSE test.
5. Run script to build app.
6. Compile app.

**Comment by rwcarlsen** [comment]
As soon as this gets merged that will by my priority.  Since merging this
doesn't actually change anything about existing fork-stork methods, I don't
think we need to expend the extra effort to have all the changes exactly
synchronized.  I will create an issue to track updates to mooseframework.org,
stork repo (deprecation), etc.

**Comment by friedmud** [comment]
I'm ok with this.

**Review by permcody — APPROVED** [review]

**Pull request merged**
